### ISSUE TITLE
Type argument entanglement in connection with & leads to wrong errors/inference

### ISSUE THREAD

**Issue by Meowtimer** [Microsoft/TypeScript#20948]
**TypeScript Version:**  2.7.0-dev.20171230

**Code**

```ts
interface RuntimeType<T> {}
interface F<T> {}
interface W<S> {}
type D<S> = (s: S) => S;

interface Z<S> {
  s: S; // commenting this also helps
}

type And<A, B> = A & B;

interface X<S, C, M> {
  a<S>(fields: { [K in keyof S]: F<S[K]> }): X<S, D<S>, M> & W<S>;
  c(): X<S, C & (() => S), M>; // triggers bug
  //b(): X<S, And<C, () => S>, M>; // workaround, encapsulating &
  f(): Z<S> & C; // commenting this also helps
}

function Z<S, C, M>(f: (X: X<{}, () => void, void>) => X<S, C, M>): Z<S> & C {
  return null;
}

function F<T>(type: RuntimeType<T>): F<T> {
  return null;
}

let isString: RuntimeType<string>;

const f = Z(b =>
  /* wrong error is here because the presence of c() does something */
  b.a({
    x: F(isString)
  })
);

f({
  x: 123 /* error should be here because the call to b.a specifies x to be 'string' */
});
```

**Expected behavior:**
Typescript complains about passing number for x when it should be string.

**Actual behavior:**
Typescript complains about something else.

**Comment by RyanCavanaugh** [comment]
Interfaces that don't use their generic type parameter will often cause inference problems; they need to manifest their type parameters in some way.

If you could please spend some time to try to simplify this and give descriptive names, that would help us a lot.
